lib/server/src/server.rs

```diff
@@ -474,65 +478,93 @@ impl TrowServer {
+        bearer_param_map.remove("realm");
 
-            if let Some(user) = &auth.user {
-                info!("Making request to Docker Hub as {}", &user);
-                auth_req = format!("{}&username={}", auth_req, user);
-            }
-            if let Some(pass) = &auth.pass {
-                auth_req = format!("{}&password={}", auth_req, pass);
-            }
+        let reqBuilder = cl.get(realm.as_str())
+            .query(
+                &bearer_param_map
+            );
 
-            cl.post(&auth.endpoint)
-                .header("Content-Type", "application/x-www-form-urlencoded")
-                .body(auth_req).send().await
+        let reqBuilder = match auth {
+            Some(a) => {
+                if a.user.is_some() {
+                    debug!("Auth is used with user: {:?}", a.user.as_ref().unwrap());
+                    reqBuilder.basic_auth(&a.user.as_ref().unwrap(), a.pass.as_ref())
+                } else {
+                    reqBuilder
+                }
+            },
+            None => reqBuilder
         };
 
-        let resp = match resp {
-            Ok(r) => r,
-            Err(e) => {
-                error!("Remote registry didn't repsond to auth request: {}", e);
-                return None
-            }
-        };
+        let resp = reqBuilder.send().await
+            .or_else(|e| Err(format_err!("Failed to send authenticate to {} request: {}", realm, e)))?;
 
         if !resp.status().is_success() {
-            error!("Failed to authenticate to {}", auth.endpoint);
-            return None
+            return Err(format_err!("Failed to authenticate to {}", realm));
         }
-        
-        let auth = match resp.json::<serde_json::Value>().await {
-            Ok(auth) => auth,
-            Err(e) => { 
-                error!("Failed to deserialize auth response {}", e); 
-                return None;
-            }
-        };
 
-        match auth["access_token"].as_str() {
-            Some(t) => Some(t.to_string()),
-            None => {
-                error!("Failed to find auth token in auth repsonse"); 
-                None
-            }
+        let auth_json_body = resp.json::<serde_json::Value>().await
+            .or_else(|e |Err(format_err!("Failed to deserialize auth response {}", e)))?;
+
+        let access_token= auth_json_body.get("access_token")
+            .ok_or(format_err!("Failed to find auth token in auth repsonse"))?;
+
+        debug!("access_token: {}",access_token);
+
+        access_token.as_str()
+            .map(|s| s.to_string())
+            .ok_or(format_err!("Access token string conversion failed"))
+    }
+
+    async fn get_www_authenticate_header(&self, cl: &reqwest::Client, image: &Image) -> Result<String, Error> {
+        let resp = cl.head(&image.get_manifest_url())
+            .headers(create_accept_header())
+            .send().await
+            .or_else(|e| Err(format_err!("Request for authenticate failed: {}", e)))?;
```

Comment:
Not sure if the reqwest error will include it, but it would be nice to put the URL in the error message.